### PR TITLE
Have weak_vendors ignore deprecations from outside

### PR DESCRIPTION
**Pull request by greg0ire**
phar:// and eval() can execute code that may or may not come from the vendors.

| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | not yet
| Fixed tickets | #24853 
| License       | MIT

I haven't managed to get the phar test to pass yet, but before I do, is it ok for me to commit a test phar in Symfony? I'm thinking trust issues? Although the phar is almost plain text.

~~Next, I'm stuck because it looks like symfony does not know how to load classes anymore... should I somehow load `vendor/autoload.php` from the phar or something like that?~~ solved, had nothing to do with that.

### CODE REVIEW

**Comment by greg0ire** [comment]
Wait wut? It passes on *some* versions of php??? I thought it wouldn't pass at all...

**Comment by sroze** [comment]
Nop, just that [some version of PHP are skipped for pull-requests](https://github.com/symfony/symfony/blob/master/.travis.yml#L101-L105). See the logs of the [successful ones](https://travis-ci.org/symfony/symfony/jobs/298749694).

**Comment by greg0ire** [comment]
Thanks @sroze , things make sense again thanks to you!

**Comment by greg0ire** [comment]
Ok so now the build only fails on 7.2... with deps=low

**Comment by greg0ire** [comment]
Could it be become of the `phpunit-bridge` vendor interfering? I can see it being installed : https://travis-ci.org/symfony/symfony/jobs/299336081#L942

**Comment by Soullivaneuh** [comment]
As a bug fix, it should be merged on 3.3: https://symfony.com/roadmap?version=3.3#checker

Should not be?

**Comment by greg0ire** [comment]
That's correct, let me rebase.

**Comment by greg0ire** [comment]
Tests are failing probably because 

> the script used to run tests is installing the latest version of the PHPUnit Bridge, which does not yet have this feature.

See https://github.com/symfony/symfony/pull/24867#issuecomment-343112567

I'm afraid there is nothing I can do about that. Is there hope to get this merged nonetheless?

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php [hunk 1]
```diff
@@ -75,9 +75,12 @@ public static function register($mode = 0)
                     }
                 }
             }
-            $path = realpath($path) ?: $path;
+            $realPath = realpath($path);
+            if (false === $realPath && '-' !== $path) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: I think `-` has changed on 7.2, the string is different

Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Dang, good catch! I knew this was a bit fragile but don't know what else to use, will have to give it more thought.

Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Thanks to docker I now know this could be "Standard input code"

Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Can't think of a less fragile solution though.

**Comment by greg0ire** [comment]
Not sure it's that string you were thinking about @nicolas-grekas : here is the change : https://github.com/php/php-src/commit/3ed8b7a87b5383f9ba99a0bbcea6e9fe7a070946 , and it does not change from `-`, but from `php://stdin`.

**Comment by greg0ire** [comment]
Ok so now I am sure, tried the code with:

```
docker run -it -v $PWD:/tmp/sf php:7.2-cli-alpine3.6 /bin/sh
cd /tmp/sf
curl -ssL https://phar.phpunit.de/phpunit.phar > phpunit
chmod +x phpunit
phpunit src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/
```

And then confirmed with some `var_dump` statements.

**Comment by greg0ire** [comment]
I just confirmed the tests pass on 5.5 by using

```
docker run -it -v $PWD:/tmp/sf melkorm/php-5.5-cli-with-extensions /bin/sh
php -d memory_limit=-1 /usr/local/bin/composer update
curl -ssL https://phar.phpunit.de/phpunit-4.8.9.phar>phpunit
chmod +x phpunit
./phpunit src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler
```

**Comment by greg0ire** [comment]
fabbot seems stuck, let's close and reopen

**Comment by greg0ire** [comment]
Please review again

**Comment by greg0ire** [comment]
ping @nicolas-grekas

**Comment by greg0ire** [comment]
I added a script to regenerate the phar, so that it's easier to understand or maintain.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @greg0ire.

**Pull request merged**